[BOLT] Add pre-parsed perf script support

kaadam opened the pull request:
This PR implements the functionality to read and parse a pre-paresed
perf-script profile which was made by Perf2bolt's
'--profile-format=perfscript' option.

It helps to add support for large ARM Spe end-to-end tests.

Why does the test need to have a textual format Spe profile?
- To collect an Arm Spe profile by Linux Perf, it needs to have
  an arm developer device which has Spe support.
- To decode Spe data, it also needs to have the proper version of
  Linux Perf.
  The minimum required version of Linux Perf is v6.15.

Bypassing these technical difficulties, that easier to prove
a pre-generated perfscript profile format.

How should generate this type of profile?

 1) You can use Perf2bolt itself to generate a pre-parsed perf-script profile
    in textual format.
    $ perf2bolt BINARY -p perf.data -o test.text --spe --profile-format=perfscript

 2) Perf2bolt is able to work with this type of profile:
    $ perf2bolt BINARY -o test.fdata  -p test.text --spe -ps

aaupov commented on bolt/include/bolt/Profile/DataAggregator.h:
```diff
@@ -440,6 +440,32 @@ class DataAggregator : public DataReader {
+  /// Then it prepares the ParsingBuf based on the classification and
+  /// call the proper functions like parseMMapEvents() or parseBranchEvents().
+  ///
+  /// This option is only for testing purposes.
```
Please add a couple of perf script examples and drop this line. You can find some under llvm/test/tools/llvm-profgen.

aaupov reviewed (commented):
Thank you for working on this, I think it's useful outside just testing. Please see couple of comments inline.
Additionally, we have `opts::ReadPerfEvents` with a similar purpose and similarly used in testing, can we unify the two options?

paschalis-mpeis commented:
> Additionally, we have opts::ReadPerfEvents with a similar purpose and similarly used in testing, can we unify the two options?

@aaupov that's the plan, to eventually replace my half-baked `ReadPerfEvents` flag :)

aaupov commented on bolt/lib/Profile/DataAggregator.cpp:
```diff
@@ -372,6 +379,80 @@ void DataAggregator::parsePreAggregated() {
+  Col = 0;
+  Line = 1;
+  std::string MMapEvents = "";
+  std::string BranchEvents = "";
```
Constructing these strings might be really expensive for large perf script inputs. Did you check if running `parseMMapEvents` and `parseBranchEvents` directly on the parsing buffer works? `parseMMapEvents` is supposed to parse as much lines as it (advancing Line/Col), returning with an error code when parsing error occurs. We can consume the error and invoke parseBranchEvents on the buffer.

kaadam replied:
Thanks for your feedback.
Yes, that was my first idea. However using the following `perf script --show-mmap-events -F pid,brstack ...` command by default to extract 'mmap' and 'brstack' events for `brstack` aggregation, I found sometimes the output is not exactly an ordered list. I mean an `mmap` event might appear among the branch events. That's why I moved to create a pre-selection logic before we call the proper functions. I agree it is costly. 

All in all if we can guarantee the strict order of the entries (as a prerequisites), that case we can call `parseMMapEvents` first, and when the first parse error occurs we can invoke `parseBranchEvents` on the Buffer as you suggested.

Side note: Using `parseMMapEvents` directly on the parsing buffer works correctly, the mmap function is able to handle other events like `brstack`. However if `parseBranchEvents` finds an mmap event, it will report an error. 

Currently I focused on branch event aggregation only, but I agree I need to cover other events and aggregation types.

paschalis-mpeis replied:
> Side note: Using parseMMapEvents directly on the parsing buffer works correctly, the mmap function is able to handle other events like brstack. However if parseBranchEvents finds an mmap event, it will report an error.

Maybe we could teach `parseBranchEvents` to ignore mmaps when we have this perf-script option used?

aaupov reviewed (commented):
Some more thoughts on this functionality (not blocking but please consider as follow-up items):
1. A note on constructing MMapEvents and BranchEvents strings - see inline comment.
2. In the general case we also need task events and buildid-list output. We should include that into perf script input as well.
3. Extra profile support: basic samples and mem profile.
4. If all of the above is implemented, the functionality should be unified with `parsePerfData` as the input would be the same (except coming from a pre-parsed file).

paschalis-mpeis commented on bolt/lib/Profile/DataAggregator.cpp:
```diff
@@ -115,6 +115,12 @@ cl::opt<std::string>
+    ReadPerfScript("perfscript",
+                   cl::desc("skip perf and read perf-script trace created by "
+                            "Linux perf tool with script command"),
+                   cl::ReallyHidden, cl::cat(AggregatorCategory));
```
```suggestion
                   cl::Hidden, cl::cat(AggregatorCategory));
```

I believe we could make the new flag just `Hidden` instead, so it can appear with the below?

```
 perf2bolt --help-list-hidden
```

The old flag (`ReadPerfEvents`) was never really used from the CLI, only in a single unit test (PerfSpeEvents), since it was never fully implemented. We could replace it and drop the old flag, given that we hardcode the parsing buffer to the values used by the relevant unit test?

paschalis-mpeis reviewed (commented):
Hi Adam,

Thanks for your work! A couple of high-level thoughts after a quick pass:

(1) For the new flag, my thinking was to provide a textual replacement for perf.data:
whatever perf2bolt would invoke, we run it beforehand outselves (ie `perf script ..`) and pass the result to the tool.

Are there actual limitations to using `--perfscript` with basic samples?

Why not call `parseBranchEvents` or `parseBasicEvents`, depending on `BasicAggregation`?
parseBranchEvents would only process the brstack events (ie PerfBranchSample) anyway, right?

(2) The flag is currently a toggle: if set, we treat -p/-perfdata as textual format, right?
We could fully align with llvm-profgen if we want and use -ps/-perfscript as a filename instead. Ofc we'll need to adjust the logic, likely in places like [this](https://github.com/llvm/llvm-project/blob/f75973949b0e51eb4b3852c903d08bc72bf5d459/bolt/tools/driver/llvm-bolt.cpp#L255). (not a strong opinion)

@kaadam, @aaupov please let me know of your thoughts.

kaadam committed:
Add support to read and parse pre-aggregated profile

This PR implements the functionality to read and parse a pre-paresed
perf-script profile which was made by Perf2bolt's
'--generate-perf-text-data' option.

It helps to add support for large ARM Spe end-to-end tests.

Why does the test need to have a textual format Spe profile?
- To collect an Arm Spe profile by Linux Perf, it needs to have
  an arm developer device which has Spe support.
- To decode Spe data, it also needs to have the proper version of
  Linux Perf.
  The minimum required version of Linux Perf is v6.15.

Bypassing these technical difficulties, that easier to prove
a pre-generated textual profile format.

How should generate this type of profile?

 1) You can use Perf2bolt itself to generate a pre-parsed perf-script profile
    in textual format.
    $ perf2bolt BINARY -p perf.data -o test.text --spe --generate-perf-script

 2) Perf2bolt is able to work with this type of profile:
    $ perf2bolt BINARY -o test.fdata  -p test.text --spe -perf-script

kaadam committed:
Add support to read and parse pre-aggregated profile

This PR implements the functionality to read and parse a pre-paresed
perf-script profile which was made by Perf2bolt's
'--generate-perf-text-data' option.

It helps to add support for large ARM Spe end-to-end tests.

Why does the test need to have a textual format Spe profile?
- To collect an Arm Spe profile by Linux Perf, it needs to have
  an arm developer device which has Spe support.
- To decode Spe data, it also needs to have the proper version of
  Linux Perf.
  The minimum required version of Linux Perf is v6.15.

Bypassing these technical difficulties, that easier to prove
a pre-generated textual profile format.

How should generate this type of profile?

 1) You can use Perf2bolt itself to generate a pre-parsed perf-script profile
    in textual format.
    $ perf2bolt BINARY -p perf.data -o test.text --spe --generate-perf-script

 2) Perf2bolt is able to work with this type of profile:
    $ perf2bolt BINARY -o test.fdata  -p test.text --spe -perf-script

kaadam commented:
FYI: This PR depends on #171144

kaadam committed:
Update PerfSpeEvent unittest

kaadam committed:
Update PerfSpeEvent unittest

paschalis-mpeis commented on bolt/lib/Profile/DataAggregator.cpp:
```diff
@@ -121,11 +121,16 @@ cl::opt<bool> ReadPreAggregated(
+
+cl::opt<bool> GeneratePerfTextProfile(
+    "generate-perf-text-data",
+    cl::desc("Dump perf-script jobs' output into output file"), cl::Hidden,
```
nit:
```suggestion
    cl::desc("Dump perf-script jobs' output into a file"), cl::Hidden,
```

paschalis-mpeis replied:
Not sure how this comment ended up here. 😅
It says I noted this on 11/Feb; probably not relevant now.

kaadam replied:
Yes, it's already updated. :)

kaadam committed:
Rename function and update cl option

kaadam committed:
Rename function and update cl option

kaadam committed:
Update MemoryMaps unittests

kaadam committed:
Update MemoryMaps unittests

paschalis-mpeis commented on bolt/lib/Profile/DataAggregator.cpp:
```diff
@@ -392,23 +392,109 @@ void DataAggregator::parsePreAggregated() {
+}
+
+void DataAggregator::parsePerfTextData(BinaryContext &BC) {
+  outs() << "PERF2BOLT: parsing a hybrid perf-script events...\n";
```
nit: Why is it hybrid? Maybe a textual or materialized?

kaadam replied:
Updated.

paschalis-mpeis commented on bolt/test/perf2bolt/Inputs/perf_test:
Was this file committed by mistake?

kaadam replied:
Yes, that happened accidentally. Thanks.

paschalis-mpeis reviewed (approved):
Hey Adam,

Thanks for your patch. Looks good overall (minus the `perf_test` file added).

Please do wait any further feedback from @aaupov.

kaadam committed:
Address reviewers

kaadam committed:
Address reviewers

kaadam committed:
Add unittest for parsing perfscript profile

Jinjie-Huang commented on bolt/lib/Profile/DataAggregator.cpp:
```diff
@@ -122,11 +122,11 @@ cl::opt<bool> ReadPreAggregated(
-                   cl::desc("skip perf event collection by supplying a "
-                            "perf-script output in a textual format"),
-                   cl::ReallyHidden, cl::init(""), cl::cat(AggregatorCategory));
+cl::opt<bool> ReadPerfTextProfile(
```
I guess a more user-friendly way might be to auto-detect the format, and set `ReadPerfTextProfile` true, since we already have `PerfTextMagicStr`. Alternatively, could we adopt a behavior similar to llvm-profgen, where we pass the filename directly to this option without needing the -p flag at all?

kaadam replied:
Yes, I agree. The main reason for adding the magic string to the generated file was indeed to enable better format auto-detection in the future. However, for this initial patch, I wanted to stick to the existing logic and focus solely on adding the "new" aggregation type in isolation, similar to 'pre-Aggregated' format.

Currently, the -p option is mandatory for 'perf2bolt', and the aggregation type must be explicitly defined if we alter the default aggregation. There is probably room for improvement here, changing this core behavior was out of scope for this task.
Thanks for the heads-up! That's a great thought for a follow-up patch to improve the user experience.

aaupov replied:
I agree with the request above: we can auto-sense perf-script format. Since perf script will be passed as `-p` arg, this will be handled by DataAggregator. `-pa` is used to drive pre-aggregated parsing. We can read the header to decide if it's pre-aggregated or perf-script format. You can add `-ps` (perf-script) as alias to `-pa`.
Does this sound reasonable to you?

kaadam replied:
@aaupov @Jinjie-Huang I've added the `-ps `option as an alias for `-pa` and merged the names of the 'cl::opts'. I also introduced a magic detection method to automatically determine which aggregation format to run. Could you please take another look?

Jinjie-Huang commented on bolt/lib/Profile/DataAggregator.cpp:
```diff
@@ -415,6 +415,105 @@ void DataAggregator::parsePreAggregated() {
+                     TimerGroupName, TimerGroupDesc, opts::TimeAggregator);
+  if (!Filename.empty()) {
+    ErrorOr<std::unique_ptr<MemoryBuffer>> MB =
+        MemoryBuffer::getFileOrSTDIN(Filename);
```
Maybe we shouldn't map the full file here just for the header parsing. Could we also use `getFileSlice` to fetch just the first 133 bytes for the header? In some cases I've encountered, perf script files grow to several tens of GBs, so mapping the whole file could be quite heavy.

kaadam replied:
Yes, that's good point. Updating that.

kaadam replied:
Done.

Jinjie-Huang commented:
@kaadam Thanks for the contribution! Overall, LGTM. A test case covering the full functionality would be great here.

kaadam commented:
@Jinjie-Huang Thanks for your review. Yes, a test case is definitely necessary. I had an offline discussion with @paschalis-mpeis, and I'm adding a unit test for this functionality at this stage. The main purpose of this feature is to support end-to-end testing for large binaries, and the plan is to further improve test coverage with 'perf-tests' later on.

kaadam committed:
Address reviewers 2

kaadam committed:
Address reviewers 2

aaupov commented on bolt/include/bolt/Profile/DataAggregator.h:
```diff
@@ -178,6 +178,8 @@ class DataAggregator : public DataReader {
     sys::ProcessInfo PI{};
     SmallVector<char, 256> StdoutPath{};
     SmallVector<char, 256> StderrPath{};
+    uint64_t Length{0};
```
Please add comments above these fields regarding their purpose (for perf-script parsing)

kaadam replied:
Done.

aaupov commented on bolt/include/bolt/Profile/DataAggregator.h:
```diff
@@ -471,6 +473,18 @@ class DataAggregator : public DataReader {
+  /// offset/length pairs for each event. Using this metadata, it opens only
+  /// the specific file slice associated with the required events during
+  /// the parsing phase.
+  void parsePerfTextData(BinaryContext &BC);
```
Should we use perf script consistently?
```suggestion
  void parsePerfScript(BinaryContext &BC);
```

kaadam replied:
Updated.

aaupov commented on bolt/lib/Profile/DataAggregator.cpp:
```diff
@@ -122,11 +122,11 @@ cl::opt<bool> ReadPreAggregated(
-                   cl::desc("skip perf event collection by supplying a "
-                            "perf-script output in a textual format"),
-                   cl::ReallyHidden, cl::init(""), cl::cat(AggregatorCategory));
+cl::opt<bool> ReadPerfTextProfile(
```
same here:
```suggestion
cl::opt<bool> ReadPerfScript(
```

aaupov commented on bolt/lib/Profile/DataAggregator.cpp:
```diff
@@ -122,11 +122,11 @@ cl::opt<bool> ReadPreAggregated(
+cl::opt<bool> ReadPerfTextProfile(
+    "perf-script",
+    cl::desc("skip perf event collection by reading a "
+             "pre-parsed perf-script output in a textual format"),
```
```suggestion
    cl::desc("read pre-parsed perf script output"),
```

aaupov commented on bolt/lib/Profile/DataAggregator.cpp:
```diff
@@ -122,11 +122,11 @@ cl::opt<bool> ReadPreAggregated(
+             "pre-parsed perf-script output in a textual format"),
+    cl::Hidden, cl::cat(AggregatorCategory));
 
 cl::opt<bool> GeneratePerfTextProfile(
```
Sorry for missing this earlier: we can move this under `profile-format=perf-script`, see #199465

Jinjie-Huang replied:
+1, extending `profile-format` support is indeed more elegant.

kaadam replied:
Thanks, I also missed that. Would you like me to update it with this change? Or Is that okay if I do it in separate PR?

aaupov replied:
Yes, let's update in this PR – you're adding perf script support and this falls squarely in it.

kaadam replied:
Done.

aaupov commented on bolt/include/bolt/Profile/DataAggregator.h:
```diff
@@ -471,6 +473,18 @@ class DataAggregator : public DataReader {
+  /// Perf2bolt first processes the pre-parsed profile's header to determine
+  /// offset/length pairs for each event. Using this metadata, it opens only
+  /// the specific file slice associated with the required events during
+  /// the parsing phase.
```
```suggestion
  /// Coordinate reading pre-parsed perf-script:
  /// - open file header to determine offset and length for each part,
  /// - read perf script slices.
```

kaadam replied:
Updated.

aaupov commented on bolt/lib/Profile/DataAggregator.cpp:
```diff
@@ -415,6 +415,106 @@ void DataAggregator::parsePreAggregated() {
 }
 
+std::error_code DataAggregator::parsePerfTextFileHeader() {
+  size_t LineEnd = ParsingBuf.find_first_of("\n");
```
Did you check if it makes sense to adopt `DataReader::parse*` methods? This will obviate the need to track Line/Col and improve diagnostics.

kaadam replied:
Yes, I did.  I cannot recall why I decided to track these myself.  I look into that.

kaadam replied:
Updated.

aaupov reviewed (commented):
I have one concern about the use of "event". This PR makes the term ambiguous: currently it means HW sampling event (`perf record -e <event>`), but here it's used in a sense of perf script invocations: "branch events" is fine, but "buildid events" doesn't make sense to me. Part or group is perhaps a better term. WDYT?

kaadam committed:
Address reviewers 3

kaadam committed:
Fix clang-format

kaadam committed:
Update

aaupov commented on bolt/include/bolt/Profile/DataAggregator.h:
```diff
@@ -176,14 +180,20 @@ class DataAggregator : public DataReader {
+    /// - Length: Total size of the content from the Offset.
+    /// - Offset: Position where content begins in the file.
+    uint64_t Length{0};
+    uint64_t Offset{0};
```
nit:
```suggestion
    /// Total size of the content from \p Offset.
    uint64_t Length{0};
    /// Position where content begins in the file.
    uint64_t Offset{0};
```

kaadam replied:
Done.

aaupov commented on bolt/lib/Profile/DataAggregator.cpp:
```diff
@@ -129,20 +129,12 @@ extern cl::opt<opts::ProfileFormatKind> ProfileFormat;
 extern cl::opt<std::string> SaveProfile;
 
-cl::opt<bool> ReadPreAggregated(
+cl::opt<bool> ReadPreAggOrPerfScript(
```
I'd just keep it as is. Should be obvious that both pa and ps are wired the same.

kaadam replied:
Done.

aaupov commented on bolt/lib/Profile/DataAggregator.cpp:
```diff
@@ -466,7 +465,112 @@ void DataAggregator::parsePreAggregated() {
+    if (std::error_code EC = MB.getError()) {
+      errs() << "PERF2BOLT-ERROR: cannot open " << Filename << ": "
+             << EC.message() << "\n";
+      exit(1);
```
I'm changing DataAggregator parsing to propagate errors (#200476), so please return an error here and below.

kaadam replied:
I've added return error into this function. I've also removed error message print.

aaupov reviewed (commented):
Thank you, this looks very good. I've left a couple of comments, please take a look. I also realize that #200476 might cause some churn for this PR as well – I'm OK with landing it after this, WDYT? Would also appreciate you reviewing it.

kaadam committed:
Address reviewers 4

kaadam commented on bolt/lib/Profile/DataAggregator.cpp:
```diff
@@ -531,6 +531,9 @@ std::error_code DataAggregator::parsePerfScriptFileHeader() {
     PPI->Length = Length;
     Offset = Offset + Length;
+
+    while (checkAndConsumeFS()) {
```
FYI: The file header has a 132-byte maximum length and it is padded with whitespace. If the contents are short, we need to consume any remaining whitespace at the end of the header. Unit-test extended.

aaupov commented on bolt/lib/Profile/DataAggregator.cpp:
```diff
@@ -736,11 +856,18 @@ void DataAggregator::imputeFallThroughs() {
+      exit(1);
+    }
+  } else if (opts::ReadPreAggregated) {
     parsePreAggregated();
```
Move checkInputFileMagic into `if (opts::ReadPreAggregated)`?

kaadam replied:
Done.

aaupov reviewed (approved):
LGTM with one nit

kaadam committed:
Address reviewers 5

paschalis-mpeis reviewed (approved):
No further suggestions.

Feel free to proceed with merging. Great work Adam!

Jinjie-Huang reviewed (approved):
Thanks

nico commented on bolt/unittests/Profile/CMakeLists.txt:
```diff
@@ -7,6 +7,12 @@ set(LLVM_LINK_COMPONENTS
+  # FIXME CoreTests uses `llvm::detail::TakeError(llvm::Error)`, but linking
+  #       to LLVMTestingSupport introduces a transitive dependency on the
+  #       dynamic LLVM library when LLVM_LINK_LLVM_DYLIB is ON.
+  ${LLVM_MAIN_SRC_DIR}/lib/Testing/Support/Error.cpp
```
What does "CoreTests" refer to here?

What's the path forward here?

It seems preferable to copy the parts of that file that you need over (and adjust the function name slightly) instead of just including an internal cpp file.

nico replied:
Come to think of it, what's the problem with the test depending on the LLVM library?

kaadam replied:
Hi Nico, Thanks for your feeedback!
CoreTests target refers to the core unitests in Bolt (bolt/unittests/Core/CMakeLists.txt).

I understand your point of view. I've applied an existing workaround from here:
https://github.com/llvm/llvm-project/blob/main/bolt/unittests/Core/CMakeLists.txt#L18

Fixing this underlying build system issue wasn't the main focus of this PR, but I'm completely open to discussing how we could handle this otherwise.
